src/OpenFeature/EventExecutor.cs

```diff
@@ -0,0 +1,294 @@
+        private readonly List<FeatureProviderReference> _activeSubscriptions = new List<FeatureProviderReference>();
+        private readonly SemaphoreSlim _shutdownSemaphore = new SemaphoreSlim(0);
+
+        private readonly Dictionary<ProviderEventTypes, List<EventHandlerDelegate>> _apiHandlers = new Dictionary<ProviderEventTypes, List<EventHandlerDelegate>>();
+        private readonly Dictionary<string, Dictionary<ProviderEventTypes, List<EventHandlerDelegate>>> _clientHandlers = new Dictionary<string, Dictionary<ProviderEventTypes, List<EventHandlerDelegate>>>();
+
+        public EventExecutor()
+        {
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+            this.ProcessEventAsync();
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+        }
+
+        internal void AddApiLevelHandler(ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            this._mutex.WaitOne();
+            if (!this._apiHandlers.TryGetValue(eventType, out var eventHandlers))
+            {
+                eventHandlers = new List<EventHandlerDelegate>();
+                this._apiHandlers[eventType] = eventHandlers;
+            }
+
+            eventHandlers.Add(handler);
+
+            this.EmitOnRegistration(this._defaultProvider, eventType, handler);
+            this._mutex.ReleaseMutex();
+        }
+
+        internal void RemoveApiLevelHandler(ProviderEventTypes type, EventHandlerDelegate handler)
+        {
+            this._mutex.WaitOne();
+            if (this._apiHandlers.TryGetValue(type, out var eventHandlers))
+            {
+                eventHandlers.Remove(handler);
+            }
+            this._mutex.ReleaseMutex();
+        }
+
+        internal void AddClientHandler(string client, ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            this._mutex.WaitOne();
+            // check if there is already a list of handlers for the given client and event type
+            if (!this._clientHandlers.TryGetValue(client, out var registry))
+            {
+                registry = new Dictionary<ProviderEventTypes, List<EventHandlerDelegate>>();
+                this._clientHandlers[client] = registry;
+            }
+
+            if (!this._clientHandlers[client].TryGetValue(eventType, out var eventHandlers))
+            {
+                eventHandlers = new List<EventHandlerDelegate>();
+                this._clientHandlers[client][eventType] = eventHandlers;
+            }
+
+            this._clientHandlers[client][eventType].Add(handler);
+
+            if (this._namedProviderReferences.TryGetValue(client, out var clientProviderReference))
+            {
+                this.EmitOnRegistration(clientProviderReference, eventType, handler);
+            }
```

Comment:
If the client name is not associated with a provider, then we have to associate the event with the default provider. This is due to how specifications define the behavior of named client [^1]

> If a client name is not bound to a particular provider, the client is associated with the default provider.

[^1]: https://openfeature.dev/specification/sections/flag-evaluation/#requirement-116

Reply:
thx for the reference, seems like I missed that part - will adapt accordingly